jenkins-dsl/jenkins_jira_integration.py

```diff
@@ -0,0 +1,352 @@
+        else:
+            print('   Did not find build_number: ' + buildnum + ' in the json cache. Populating...')
+            # Pull the data from the Jenkins server as we don't have it in our cache
+            build_data = BuildData(server, branch, buildnum)
+
+            # The data in our JSON cache is a pretty simple subset of what we have in BuildData for a full CI report
+            newcache_data = {}
+            newcache_data[NUMBER] = build_data.number
+            newcache_data[PREVIOUS_NUMBER] = build_data.previous_number
+            # Can't serialize sets to JSON
+            newcache_data['failures'] = list(build_data.test_failures)
+
+            log('   Got data for ' + str(newcache_data[NUMBER]) + '. Caching and setting previous build pointer to: ' + str(newcache_data[PREVIOUS_NUMBER]))
+            if 'failures' in newcache_data:
+                log('   build number: ' + str(newcache_data[NUMBER]) + ' with failure count: ' + str(len(newcache_data['failures'])))
+
+            # JSON comes back out as a string; need to follow suit here so the cache is all the same data type
+            cached_data[str(build_data.number)] = newcache_data
+
+            # cast to str should be redundant but it's coming out as an int. Not worth arguing with it.
+            buildnum = str(newcache_data[PREVIOUS_NUMBER])
+
+    log('   Hit limit of builds. Saving JSON.')
+    with open(json_file, 'w', encoding='utf-8') as outfile:
+        log('Updating the cache with recently queried results')
+        json.dump(cached_data, outfile)
+
+    return cached_data
+
+
+def get_test_failure_details(jira: JIRA, test_name: str, data: Dict) -> tuple[int, int, str]:
+    """
+    :param data: Dict representing JSON cached data
+    :return: failures, total, test failure JIRA url if found, link to test board if not
+    """
+    fail_count = 0
+    total = 0
+
+    for build in data:
+        build_data = data[build]
+        total += 1
+        if test_name in build_data['failures']:
+            fail_count += 1
+
+    # See if we have a ticket for this failure already; pull out last 2 tokens from the . delimited test FQN
+    tokens = test_name.split('.')
+
+    # Strip out anything in [] dynamically created in the last token; JQL is not a fan of braces.
+    # More specifically, attempting to escape them is getting dropped by the jira python API and I'm
+    # not really interested in figuring out why; the class + name should be sufficient to find a
+    # failure Jira if it exists
+    subtoken = tokens[-1].split('[')
+    short_name = tokens[-2] + '.' + subtoken[0]
+
+    # If we have a runtime configured test name with braces in it, we search for the root test class name only since JQL
+    # is not fond of braces
+    query = 'project = CASSANDRA and resolution = unresolved and summary ~ "' + short_name + '"'
+    log('About to query via the following JQL: ' + query)
+    try:
+        has_jira = jira.search_issues(query)
```

Comment:
does this do retries?

Reply:
As in, mutate the JQL to try with a different form, or gracefully deal with the fact that sometimes Jira won't reply?

In the event Jira doesn't reply, the link it generates will be to the parent kanban board so theoretically a dev would click on that, see the ticket, and not create a new one. This is assuming folks even click these links... 😢